src/Features/Core/Portable/FileHeaders/AbstractFileHeaderHelper.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            SingleLineCommentTriviaKind = syntaxKinds.SingleLineCommentTrivia;
+            MultiLineCommentTriviaKind = syntaxKinds.MultiLineCommentTrivia;
+            WhitespaceTriviaKind = syntaxKinds.WhitespaceTrivia;
+            EndOfLineTriviaKind = syntaxKinds.EndOfLineTrivia;
+        }
+
+        public abstract string CommentPrefix { get; }
+
+        private int SingleLineCommentTriviaKind { get; }
+        private int MultiLineCommentTriviaKind { get; }
+        private int WhitespaceTriviaKind { get; }
+        private int EndOfLineTriviaKind { get; }
+
+        protected abstract string GetTextContextOfComment(SyntaxTrivia commentTrivia);
+
+        public FileHeader ParseFileHeader(SyntaxNode root)
+        {
+            var firstToken = root.GetFirstToken(includeZeroWidth: true);
+            var firstNonWhitespaceTrivia = IndexOfFirstNonWhitespaceTrivia(firstToken.LeadingTrivia, true);
+
+            if (firstNonWhitespaceTrivia == -1)
+            {
+                return FileHeader.MissingFileHeader(0);
+            }
+
+            var sb = StringBuilderPool.Allocate();
+            var endOfLineCount = 0;
+            var done = false;
+            var missingHeaderOffset = 0;
+            var fileHeaderStart = int.MaxValue;
+            var fileHeaderEnd = int.MinValue;
+
+            for (var i = firstNonWhitespaceTrivia; !done && (i < firstToken.LeadingTrivia.Count); i++)
+            {
+                var trivia = firstToken.LeadingTrivia[i];
+
+                if (trivia.RawKind == WhitespaceTriviaKind)
+                {
+                    endOfLineCount = 0;
+                }
+                else if (trivia.RawKind == SingleLineCommentTriviaKind)
+                {
+                    endOfLineCount = 0;
+
+                    var commentText = GetTextContextOfComment(trivia);
+
+                    fileHeaderStart = Math.Min(trivia.FullSpan.Start, fileHeaderStart);
+                    fileHeaderEnd = trivia.FullSpan.End;
+
+                    sb.AppendLine(commentText.Trim());
+                }
+                else if (trivia.RawKind == MultiLineCommentTriviaKind)
+                {
+                    // only process a MultiLineCommentTrivia if no SingleLineCommentTrivia have been processed
+                    if (sb.Length == 0)
+                    {
+                        var commentText = GetTextContextOfComment(trivia);
+                        var triviaStringParts = commentText.Trim().Replace("\r\n", "\n").Split('\n');
+
+                        foreach (var part in triviaStringParts)
+                        {
+                            var trimmedPart = part.TrimStart(' ', '*');
+                            sb.AppendLine(trimmedPart);
+                        }
+
+                        fileHeaderStart = trivia.FullSpan.Start;
+                        fileHeaderEnd = trivia.FullSpan.End;
+                    }
+
+                    done = true;
```

Comment:
i don't get the benefit of a 'done' flag instead of just literally `break`ing in these cases.

Reply:
Switched to `break`